### ISSUE TITLE
manual_backward and .backward() have different behaviour.

### ISSUE THREAD

**Issue by roedoejet** [Lightning-AI/lightning#18740]
### Bug description

I expected manual_backward and .backward to perform backward propagation in the same way, but when I use self.manual_backward it results in a number of unused parameters. If I use .backward then the problem doesn't occur.

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
def training_step(self, batch, batch_idx):
        x, y, _, y_mel = batch
        y = y.unsqueeze(1)
        # x.size() & y_mel.size() = [batch_size, n_mels=80, n_frames=32]
        # y.size() = [batch_size, segment_size=8192]
        optim_g, optim_d = self.optimizers()
        scheduler_g, scheduler_d = self.lr_schedulers()
        # generate waveform
        if self.config.model.istft_layer:
            mag, phase = self(x)
            generated_wav = self.inverse_spectral_transform(
                mag * torch.exp(phase * 1j)
            ).unsqueeze(-2)
        else:
            generated_wav = self(x)
       
        # create mel
        generated_mel_spec = dynamic_range_compression_torch(
            self.spectral_transform(generated_wav).squeeze(1)[:, :, 1:]
        )
        # train discriminators
        optim_d.zero_grad()
        # MPD
        y_df_hat_r, y_df_hat_g, _, _ = self.mpd(y, generated_wav.detach())
        if self.use_gradient_penalty:
            gp_f = self.compute_gradient_penalty(y.data, generated_wav.detach().data, self.mpd)
        else:
            gp_f = None
        loss_disc_f, _, _ = self.discriminator_loss(y_df_hat_r, y_df_hat_g, gp=gp_f)
        self.log("training/disc/mpd_loss", loss_disc_f, prog_bar=False)
        # MSD
        y_ds_hat_r, y_ds_hat_g, _, _ = self.msd(y, generated_wav.detach())
        loss_disc_s, _, _ = self.discriminator_loss(y_ds_hat_r, y_ds_hat_g, gp=gp_s)
        self.log("training/disc/msd_loss", loss_disc_s, prog_bar=False)
        # calculate loss
        disc_loss_total = loss_disc_s + loss_disc_f
        # manual optimization because Pytorch Lightning 2.0+ doesn't handle automatic optimization for multiple optimizers
        # this works
        disc_loss_total.backward()
        # this does not
        # self.manual_backward(disc_loss_total
        optim_d.step()
        scheduler_d.step()
        # log discriminator loss
        self.log("training/disc/d_loss_total", disc_loss_total, prog_bar=False)
            
        # train generator
        optim_g.zero_grad()
        # calculate loss
        _, y_df_hat_g, fmap_f_r, fmap_f_g = self.mpd(y, generated_wav)
        _, y_ds_hat_g, fmap_s_r, fmap_s_g = self.msd(y, generated_wav)
        loss_fm_f = self.feature_loss(fmap_f_r, fmap_f_g)
        loss_fm_s = self.feature_loss(fmap_s_r, fmap_s_g)
        loss_gen_f, _ = self.generator_loss(
            y_df_hat_g, gp=self.use_gradient_penalty
        )
        loss_gen_s, _ = self.generator_loss(
            y_ds_hat_g, gp=self.use_gradient_penalty
        )
        self.log("training/gen/loss_fmap_f", loss_fm_f, prog_bar=False)
        self.log("training/gen/loss_fmap_s", loss_fm_s, prog_bar=False)
        self.log("training/gen/loss_gen_f", loss_gen_f, prog_bar=False)
        self.log("training/gen/loss_gen_s", loss_gen_s, prog_bar=False)
        loss_mel = F.l1_loss(y_mel, generated_mel_spec) * 45
        gen_loss_total = loss_gen_s + loss_gen_f + loss_fm_s + loss_fm_f + loss_mel
        # manual optimization because Pytorch Lightning 2.0+ doesn't handle automatic optimization for multiple optimizers
        gen_loss_total.backward()
        optim_g.step()
        scheduler_g.step()
        # log generator loss
        self.log("training/gen/gen_loss_total", gen_loss_total, prog_bar=True)
        self.log("training/gen/mel_spec_error", loss_mel / 45, prog_bar=False)


I caught this by adding an `on_after_backward` method. When I use `self.manual_backward(disc_loss_total)` or `self.manual_backward(gen_loss_total)` then I get a bunch of parameters with `p.grad == None` but when I use `disc_loss_total.backward()` everything works fine.
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
	- GPU:
		- Tesla V100-SXM2-16GB
	- available:         True
	- version:           11.7
* Lightning:
	- lightning:         2.0.4
	- lightning-cloud:   0.5.39
	- lightning-utilities: 0.9.0
	- pytorch-lightning: 2.0.9.post0
	- torch:             2.0.1+cu117
	- torchaudio:        2.0.2+cu117
	- torchmetrics:      1.2.0
* Packages:
	- absl-py:           2.0.0
	- aiohttp:           3.8.5
	- aiosignal:         1.3.1
	- aniso8601:         9.0.1
	- annotated-types:   0.5.0
	- anyio:             3.7.1
	- anytree:           2.9.0
	- arrow:             1.3.0
	- async-timeout:     4.0.3
	- attrs:             23.1.0
	- audioread:         3.0.1
	- beautifulsoup4:    4.12.2
	- bidict:            0.22.1
	- black:             22.12.0
	- blessed:           1.20.0
	- cachetools:        5.3.1
	- certifi:           2023.7.22
	- cffi:              1.16.0
	- cfgv:              3.4.0
	- charset-normalizer: 3.3.0
	- click:             8.1.7
	- clipdetect:        0.1.3
	- cmake:             3.27.6
	- colorama:          0.4.6
	- coloredlogs:       14.0
	- contourpy:         1.1.1
	- croniter:          1.3.15
	- cycler:            0.12.0
	- cython:            3.0.3
	- dateutils:         0.6.12
	- decorator:         5.1.1
	- deepdiff:          6.6.0
	- distlib:           0.3.7
	- dnspython:         2.3.0
	- editdistance:      0.6.2
	- einops:            0.5.0
	- et-xmlfile:        1.1.0
	- eventlet:          0.33.3
	- everyvoice:        0.1.20231005
	- exceptiongroup:    1.1.3
	- fastapi:           0.103.2
	- filelock:          3.12.4
	- flake8:            6.1.0
	- flask:             2.2.5
	- flask-cors:        4.0.0
	- flask-restful:     0.3.10
	- flask-socketio:    5.3.6
	- flask-talisman:    1.1.0
	- fonttools:         4.43.0
	- frozenlist:        1.4.0
	- fsspec:            2023.9.2
	- g2p:               1.1.20230822
	- gitlint-core:      0.19.1
	- google-auth:       2.23.2
	- google-auth-oauthlib: 1.0.0
	- greenlet:          3.0.0
	- grpcio:            1.59.0
	- h11:               0.14.0
	- humanfriendly:     10.0
	- identify:          2.5.30
	- idna:              3.4
	- importlib-metadata: 6.8.0
	- iniconfig:         2.0.0
	- inquirer:          3.1.3
	- isort:             5.12.0
	- itsdangerous:      2.1.2
	- jinja2:            3.1.2
	- joblib:            1.3.2
	- jsonschema:        4.19.1
	- jsonschema-specifications: 2023.7.1
	- kiwisolver:        1.4.5
	- librosa:           0.9.2
	- lightning:         2.0.4
	- lightning-cloud:   0.5.39
	- lightning-utilities: 0.9.0
	- lit:               17.0.2
	- llvmlite:          0.41.0
	- loguru:            0.6.0
	- markdown:          3.4.4
	- markdown-it-py:    3.0.0
	- markupsafe:        2.1.3
	- matplotlib:        3.6.0
	- mccabe:            0.7.0
	- mdurl:             0.1.2
	- merge-args:        0.1.5
	- mpmath:            1.3.0
	- multidict:         6.0.4
	- munkres:           1.1.4
	- mypy:              1.5.1
	- mypy-extensions:   1.0.0
	- networkx:          2.8.4
	- nltk:              3.7
	- nodeenv:           1.8.0
	- numba:             0.58.0
	- numpy:             1.25.2
	- oauthlib:          3.2.2
	- openpyxl:          3.1.2
	- ordered-set:       4.1.0
	- packaging:         23.2
	- pandas:            1.4.4
	- panphon:           0.20.0
	- pathspec:          0.11.2
	- pillow:            10.0.1
	- pip:               23.2.1
	- platformdirs:      3.11.0
	- pluggy:            1.3.0
	- pooch:             1.7.0
	- pre-commit:        3.4.0
	- prompt-toolkit:    3.0.39
	- protobuf:          4.24.4
	- psutil:            5.9.5
	- pyasn1:            0.5.0
	- pyasn1-modules:    0.3.0
	- pycodestyle:       2.11.0
	- pycountry:         22.3.5
	- pycparser:         2.21
	- pydantic:          2.4.2
	- pydantic-core:     2.10.1
	- pyflakes:          3.1.0
	- pygments:          2.16.1
	- pyjwt:             2.8.0
	- pympi-ling:        1.70.2
	- pyparsing:         3.1.1
	- pysdtw:            0.0.5
	- pytest:            7.4.2
	- python-dateutil:   2.8.2
	- python-editor:     1.0.4
	- python-engineio:   4.7.1
	- python-multipart:  0.0.6
	- python-socketio:   5.9.0
	- pytorch-lightning: 2.0.9.post0
	- pytz:              2023.3.post1
	- pyworld:           0.3.4
	- pyyaml:            6.0.1
	- questionary:       1.10.0
	- readchar:          4.0.5
	- referencing:       0.30.2
	- regex:             2023.10.3
	- requests:          2.31.0
	- requests-oauthlib: 1.3.1
	- resampy:           0.4.2
	- rich:              13.6.0
	- rpds-py:           0.10.4
	- rsa:               4.9
	- scikit-learn:      1.3.1
	- scipy:             1.11.3
	- setuptools:        59.5.0
	- sh:                2.0.6
	- shellingham:       1.5.3
	- simple-term-menu:  1.5.2
	- simple-websocket:  1.0.0
	- six:               1.16.0
	- sniffio:           1.3.0
	- soundfile:         0.12.1
	- soupsieve:         2.5
	- starlette:         0.27.0
	- starsessions:      1.3.0
	- sympy:             1.12
	- tabulate:          0.8.10
	- tensorboard:       2.14.1
	- tensorboard-data-server: 0.7.1
	- text-unidecode:    1.3
	- threadpoolctl:     3.2.0
	- tomli:             2.0.1
	- torch:             2.0.1+cu117
	- torchaudio:        2.0.2+cu117
	- torchmetrics:      1.2.0
	- tqdm:              4.66.1
	- traitlets:         5.11.2
	- triton:            2.0.0
	- typer:             0.9.0
	- types-python-dateutil: 2.8.19.14
	- types-pyyaml:      6.0.12.12
	- types-requests:    2.31.0.8
	- types-setuptools:  68.2.0.0
	- types-tabulate:    0.8.11
	- typing-extensions: 4.8.0
	- unicodecsv:        0.14.1
	- urllib3:           2.0.6
	- uvicorn:           0.23.2
	- virtualenv:        20.24.5
	- wcwidth:           0.2.8
	- websocket-client:  1.6.3
	- websockets:        11.0.3
	- werkzeug:          2.2.3
	- wheel:             0.41.2
	- wsproto:           1.2.0
	- yarl:              1.9.2
	- zipp:              3.17.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.9.18
	- release:           4.15.0-204-generic
	- version:           #215-Ubuntu SMP Fri Jan 20 18:24:59 UTC 2023

</details>

**Comment by awaelchli** [comment]
Hey @roedoejet 
Is this reproducible in our GAN examples?

**Comment by roedoejet** [comment]
> Hey @roedoejet 
> Is this reproducible in our GAN examples?
> https://github.com/Lightning-AI/lightning/blob/master/examples/pytorch/domain_templates/generative_adversarial_net.py

I'm away now until next week, but I will give it a shot and post and update here. Thanks.

**Comment by awaelchli** [comment]
@roedoejet Thanks for looking at it. Due to priorities, I won't have the bandwidth to search for the bug. If you find a way to reproduce this in a code example we can study, that would be invaluable.

**Comment by deependujha** [comment]
I dug into this since it's been open a while without a runnable reproducer, and I think the difference the original report saw is a measurement artifact rather than a real difference in what gets computed.

The key detail: `on_after_backward` only runs when backward goes through Lightning, that's both `automatic optimization` and `self.manual_backward(...)`. A plain `loss.backward()` called by hand inside `training_step` skips Lightning's backward path entirely, so `on_after_backward` never fires for it.

In the original code, the generator output is detached for the discriminator loss (`generated_wav.detach()`), and there's no `toggle_optimizer`/`untoggle_optimizer`. So after the discriminator backward, the generator params genuinely have no gradient yet that's correct, not a bug. With `manual_backward`, `on_after_backward` runs and surfaces those `None` grads, so it looks like "unused parameters." With the hand-rolled `loss.backward()`, `on_after_backward` never runs, so those same `None` grads were simply never observed, which is why that path looked fine. The gradients themselves are identical either way; only one path was showing them.

I reproduced this in a GAN-style manual-optimization loop (two sub-models, two optimizers, detached generator output) and compared `manual_backward` against a hand-rolled `loss.backward()`, checking grads directly inside `training_step` instead of via the hook. The gradient state is identical at every step: generator params have no grad after the discriminator backward, everything has a grad after the generator backward, same in both cases. I also confirmed `manual_backward` works as expected under DDP (multi-GPU), with `on_after_backward` firing on each rank.

One note: `manual_backward` is the call you want in manual optimization, a hand-rolled `loss.backward()` skips Lightning's backward handling (and under DDP, the gradient synchronization step), so it isn't really an equivalent option even though the single-device gradients happen to match.

https://github.com/Lightning-AI/pytorch-lightning/blob/fe6b1cc4e80ae0396e2e404c16e6b6968ad5437e/src/lightning/pytorch/overrides/distributed.py#L42-L59

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader, Dataset

import lightning as L
from lightning.pytorch import Trainer

# Flip to compare the two backward paths.
USE_MANUAL_BACKWARD = False
print(f"{USE_MANUAL_BACKWARD=}")


class Generator(nn.Module):
    def __init__(self):
        super().__init__()
        self.net = nn.Sequential(nn.Linear(16, 32), nn.ReLU(), nn.Linear(32, 32))

    def forward(self, x):
        return self.net(x)


class Discriminator(nn.Module):
    def __init__(self):
        super().__init__()
        self.net = nn.Sequential(nn.Linear(32, 16), nn.ReLU(), nn.Linear(16, 1))

    def forward(self, x):
        return self.net(x)


def grad_status(module, prefix):
    return {f"{prefix}.{n}": (p.grad is not None) for n, p in module.named_parameters()}


class GAN(L.LightningModule):
    def __init__(self, use_manual_backward: bool):
        super().__init__()
        self.use_manual_backward = use_manual_backward
        self.automatic_optimization = False
        self.generator = Generator()
        self.discriminator = Discriminator()
        self._report = {}

    def configure_optimizers(self):
        opt_g = torch.optim.Adam(self.generator.parameters(), lr=1e-3)
        opt_d = torch.optim.Adam(self.discriminator.parameters(), lr=1e-3)
        return [opt_g, opt_d]

    def _backward(self, loss):
        if self.use_manual_backward:
            self.manual_backward(loss)
        else:
            # Hand-rolled .backward() inside a manual-optimization step.
            # NOTE: under DDP this bypasses prepare_for_backward and is not a
            # supported pattern; included only to compare grad coverage on CPU.
            loss.backward()

    def training_step(self, batch, batch_idx):
        x, real = batch
        opt_g, opt_d = self.optimizers()
        generated = self.generator(x)

        # Discriminator step (generator output detached)
        self.untoggle_optimizer(opt_g)
        self.toggle_optimizer(opt_d)
        opt_d.zero_grad()
        real_score = self.discriminator(real)
        fake_score = self.discriminator(generated.detach())
        d_loss = (
            F.binary_cross_entropy_with_logits(real_score, torch.ones_like(real_score))
            + F.binary_cross_entropy_with_logits(fake_score, torch.zeros_like(fake_score))
        )
        self._backward(d_loss)
        d_grads = {**grad_status(self.discriminator, "disc"), **grad_status(self.generator, "gen")}
        opt_d.step()
        self.untoggle_optimizer(opt_d)

        # Generator step
        self.toggle_optimizer(opt_g)
        opt_g.zero_grad()
        fake_score_for_g = self.discriminator(generated)
        g_loss = F.binary_cross_entropy_with_logits(fake_score_for_g, torch.ones_like(fake_score_for_g))
        self._backward(g_loss)
        g_grads = {**grad_status(self.generator, "gen"), **grad_status(self.discriminator, "disc")}
        opt_g.step()
        self.untoggle_optimizer(opt_g)

        if batch_idx == 0:
            self._report = {"after_d_backward": d_grads, "after_g_backward": g_grads}

    def on_after_backward(self):
        print(f"[rank {self.global_rank}] on_after_backward fired")

    def on_train_batch_end(self, outputs, batch, batch_idx):
        if batch_idx == 0 and self.global_rank == 0:
            mode = "manual_backward" if self.use_manual_backward else "loss.backward()"
            print(f"\n=== mode: {mode} ===")
            for phase, grads in self._report.items():
                print(f"\n{phase}:")
                for name, has in grads.items():
                    print(f"  {name}: {'HAS GRAD' if has else '** NO GRAD **'}")


class RandomDataset(Dataset):
    def __init__(self, length=64):
        self.x = torch.randn(length, 16)
        self.y = torch.randn(length, 32)

    def __getitem__(self, i):
        return self.x[i], self.y[i]

    def __len__(self):
        return len(self.x)


def main():
    torch.manual_seed(0)
    loader = DataLoader(RandomDataset(), batch_size=16)
    model = GAN(use_manual_backward=USE_MANUAL_BACKWARD)
    trainer = Trainer(
        accelerator="cpu",  # CPU so the hand-rolled .backward() comparison is valid
        devices=1,
        max_epochs=1,
        enable_checkpointing=False,
        enable_progress_bar=False,
        enable_model_summary=False,
        logger=False,
    )
    trainer.fit(model, loader)


if __name__ == "__main__":
    main()
```

<img width="1449" height="609" alt="Image" src="https://github.com/user-attachments/assets/8b7eb117-fcd9-41f4-8bd2-3872ccaf8aa5" />

based on the image, it should be very clear that using `loss.backward()` won't even call `on_after_backward()`, whereas `self.manual_backward(loss)`, ensures callbacks are called, and correctly shows parameters that have gradient to be `None`. `toggle/untoggle` is the right way to manage which optimizer's params are active per step.

---

No runnable reproducer in ~3 years, and from the investigation above this looks like a measurement artifact: `loss.backward()` doesn't fire `on_after_backward`, while `manual_backward` does and surfaces the params with no grad at that point, but the gradients themselves are identical. Seems safe to close; happy to revisit if anyone has a minimal repro showing an actual gradient difference.
